File: OnlyCue/LTC/SMPTEFramerate.swift
import Foundation

/// The SMPTE timecode rates OnlyCue's LTC generator supports (epic #33).
/// `23.976` / `29.97` / `59.94` and pulldown are out of scope for v1.
/// The `rawValue` is the stable on-disk token used when the project framerate
/// is persisted in `.cuelist` (a later leaf) — don't change one without a
/// migration.
enum SMPTEFramerate: String, Codable, CaseIterable, Identifiable, Sendable {
    case fps24 = "24"
    case fps25 = "25"
    case fps30 = "30"
    case fps30drop = "30df"

    var id: String { rawValue }

    /// Frames per second, nominal. For `fps30drop` the *physical* rate in real
    /// broadcast hardware is 29.97; v1 treats it as a 30 fps timeline whose
    /// timecode *labels* follow the drop-frame counting rule (see `Timecode`),
    /// which is the simplification the epic asks for (29.97 itself is excluded).
    var framesPerSecond: Int {
        switch self {
        case .fps24: return 24
        case .fps25: return 25
        case .fps30, .fps30drop: return 30
        }
    }

    var isDropFrame: Bool { self == .fps30drop }

    var displayName: String {
        switch self {
        case .fps24: return "24 fps"
        case .fps25: return "25 fps"
        case .fps30: return "30 fps (non-drop)"
        case .fps30drop: return "30 fps (drop-frame)"
        }
    }
}

File: windows/OnlyCue.Core/SmpteFramerate.cs
namespace OnlyCue.Core;

/// <summary>
/// The SMPTE frame rates OnlyCue supports. Mirrors the Swift
/// <c>SMPTEFramerate</c> (<c>OnlyCue/LTC/SMPTEFramerate.swift</c>); the
/// <see cref="RawValue"/> strings are the cross-platform contract keys used by
/// <c>golden/timecode-v1.json</c>.
/// </summary>
public enum SmpteFramerate
{
    Fps24,
    Fps25,
    Fps30,
    Fps30Drop
}

public static class SmpteFramerateExtensions
{
    /// <summary>The nominal frame count per second — 30 for drop-frame, which
    /// labels frames at 30 and drops numbers rather than changing the base.</summary>
    public static int FramesPerSecond(this SmpteFramerate rate) => rate switch
    {
        SmpteFramerate.Fps24 => 24,
        SmpteFramerate.Fps25 => 25,
        SmpteFramerate.Fps30 => 30,
        SmpteFramerate.Fps30Drop => 30,
        _ => throw new ArgumentOutOfRangeException(nameof(rate))
    };

    public static bool IsDropFrame(this SmpteFramerate rate) => rate == SmpteFramerate.Fps30Drop;

    /// <summary>The contract key as written in the golden vectors.</summary>
    public static string RawValue(this SmpteFramerate rate) => rate switch
    {
        SmpteFramerate.Fps24 => "24",
        SmpteFramerate.Fps25 => "25",
        SmpteFramerate.Fps30 => "30",
        SmpteFramerate.Fps30Drop => "30df",
        _ => throw new ArgumentOutOfRangeException(nameof(rate))
    };

    /// <summary>
    /// The rate with this nominal frames-per-second and drop-frame flag, or
    /// <c>null</c> if there is none (drop-frame only exists at 30 fps; only
    /// 24 / 25 / 30 are supported). Used when recovering a rate from a decoded
    /// LTC signal: the magnitude comes from the measured bit period, the
    /// drop-frame bit from the frame itself.
    /// </summary>
    public static SmpteFramerate? Matching(int framesPerSecond, bool isDropFrame)
    {
        if (isDropFrame)
        {
            return framesPerSecond == 30 ? SmpteFramerate.Fps30Drop : null;
        }

        return framesPerSecond switch
        {
            24 => SmpteFramerate.Fps24,
            25 => SmpteFramerate.Fps25,
            30 => SmpteFramerate.Fps30,
            _ => null
        };
    }

    public static SmpteFramerate FromRawValue(string rawValue) => rawValue switch
    {
        "24" => SmpteFramerate.Fps24,
        "25" => SmpteFramerate.Fps25,
        "30" => SmpteFramerate.Fps30,
        "30df" => SmpteFramerate.Fps30Drop,
        _ => throw new ArgumentException($"unknown framerate '{rawValue}'", nameof(rawValue))
    };
}
